src/methods.jl

```diff
@@ -10,318 +19,453 @@ derivatives.
-    end
-    eps > 0 || throw(ArgumentError("eps must be positive, got $eps"))
-    bound > 0 || throw(ArgumentError("bound must be positive, got $bound"))
-    0 <= adapt < 20 - m.p || throw(ArgumentError("can't perform $adapt adaptation steps"))
-
-    # The below calculation can fail for `m.q == 0`, because then `C₂` may be zero. We
-    # therefore handle this edge case here.
-    m.q == 0 && return m, f(x)
-
-    p = m.p
+function estimate_step(
+    m::FiniteDifferenceMethod,
+    f::Function,
+    x::T;
+    factor::Int=1,
+    max_step=convert(T, 0.1)
+) where T<:AbstractFloat
+    p = length(m.coefs)
     q = m.q
-    grid = m.grid
-    coefs = m.coefs
-
-    # Adaptively compute the bound on the function and derivative values, if applicable.
-    if adapt > 0
-        newm = (M.name.wrapper)(p + 1, p)
-        dfdx = fdm(
-            newm,
-            f,
-            x;
-            condition=condition,
-            eps=eps,
-            bound=bound,
-            max_step=max_step,
-            adapt=(adapt - 1),
-        )
-        bound = _estimate_bound(dfdx, condition)
-    end
+    f_x = float(f(x))
+
+    # Estimate the bound and round-off error.
+    ε = add_tiny(maximum(eps.(f_x))) * factor
+    M = add_tiny(m.bound_estimator(f, x))
 
     # Set the step size by minimising an upper bound on the error of the estimate.
-    C₁ = eps * sum(abs, coefs)
-    C₂ = bound * sum(n->abs(coefs[n] * grid[n]^p), eachindex(coefs)) / factorial(p)
-    ĥ = convert(T, min((q / (p - q) * C₁ / C₂)^(1 / p), max_step))
+    C₁ = ε * sum(abs, m.coefs)
+    C₂ = M * sum(n -> abs(m.coefs[n] * m.grid[n]^p), eachindex(m.coefs)) / factorial(p)
+    h = convert(T, min((q / (p - q) * C₁ / C₂)^(1 / p), max_step))
 
     # Estimate the accuracy of the method.
-    accuracy = ĥ^(-q) * C₁ + ĥ^(p - q) * C₂
+    accuracy = h^(-q) * C₁ + h^(p - q) * C₂
 
-    # Estimate the value of the derivative.
-    dfdx = sum(i -> convert(T, coefs[i]) * f(T(x + ĥ * grid[i])), eachindex(grid)) / ĥ^q
+    return h, accuracy
+end
 
-    m.history.eps = eps
-    m.history.bound = bound
-    m.history.step = ĥ
-    m.history.accuracy = accuracy
+for direction in [:forward, :central, :backward]
+    fdm_fun = Symbol(direction, "_fdm")
+    grid_fun = Symbol("_", direction, "_grid")
+    @eval begin function $fdm_fun(
+            p::Int,
+            q::Int;
+            adapt::Int=1,
+            condition::Int=DEFAULT_CONDITION,
+            geom::Bool=false
+        )
+            _check_p_q(p, q)
+            grid = collect($grid_fun(p))
+            geom && (grid = _exponentiate_grid(grid))
+            coefs = _coefs(grid, q)
+            return FiniteDifferenceMethod(
+                grid,
+                q,
+                coefs,
+                _make_adaptive_bound_estimator(
+                    $fdm_fun,
+                    p,
+                    adapt,
+                    condition,
+                    geom=geom
+                )
```

Comment:
This just fits in the line limit

```suggestion
                _make_adaptive_bound_estimator($fdm_fun, p, adapt, condition, geom=geom),
```